AST without autocorrection

cryogenian opened the pull request:
Tests say that it works. 

Project has four parsers `String ->  Either _ [Token]`, `[Token] -> Either _ [Value]`, `[Value] -> Either _ [PredicateAndLabel]` and `[PredicateAndLabel] -> Either _ [SearchTerm]`, they are chained in `Text.SlamSearch.Parser` module. 

Could you, please, provide some examples of autocorrection? Is it just `. -> ..` ``... -> ..` or something more complicated?

jdegoes committed:
Initial commit

jdegoes committed:
Parser without error correction

switched to mandragora-bucket

added version to bower.json

removed Utils in favour of modified Text.Parsing

switched to purescript-exception

one constructor in SearchTermSimple, added getPredicate

Value -> TextVal, isVal -> isTextual, isValue -> isTextVal

renamed somethingP -> somethingPredicate, SimplePredicate -> Predicate

refined export and import

SearchQuery added

updated mandragora-bucket, travis should be able to build it now

updated psc

version updated

printer added

tests

removed unneeded utils

updated mandragora-bucket to max inactivity time

cryogenian committed:
Parser without error correction

jdegoes commented on gulpfile.js:
```diff
@@ -0,0 +1,249 @@
+var gulp = require("gulp"),
```
There's quite a lot of boilerplate in all these SlamData projects (that is, the gulp files look almost the same). Would it make sense to factor this "skeleton" out into a gulp plugin, for example?

cryogenian replied:
ok

garyb replied:
I had started working on some changes to the `gulpfile` in the `slamdata` project - I had some issues with getting it to build, maybe just a Windows thing as I've not tried it on a Mac yet... but aside from that, I switched to `psc-make` with `browserify` too, as it can make for much faster incremental compiles. I can break that out into its own project perhaps.

cryogenian replied:
I've made project with name `cryogenian/mandragora-bucket` with all gulp tasks and deps. Packaged via npm.

jdegoes commented on package.json:
```diff
@@ -0,0 +1,48 @@
+{
+    "name": "purescript-search",
+    "version": "0.0.0",
```
Let's make this version 0.1.0 so we can semantically version it.

jdegoes commented on src/Text/SlamSearch/Parser.purs:
```diff
@@ -0,0 +1,17 @@
+import Text.SlamSearch.Parser.Tokens
+import Text.SlamSearch.Parser.Values
+
+parseSearchQuery :: String -> Either ParseError [SearchTerm]
```
Like most search grammars, whitespace represents implicit conjunction. That is, "foo bar" means to search for things which have both "foo" AND "bar". While we can encode this top-level conjunction in an array, it may be better to encode it using an ADT. E.g. `SearchAND foo bar`. The reason is that it paves the way to supporting `OR` in the future: "foo OR bar", which would be parsed into `SearchOR foo bar`.

jdegoes commented on src/Text/SlamSearch/Parser/Term.purs:
```diff
@@ -0,0 +1,202 @@
+
+data SearchTermSimple =
+  SimplePredicate SimplePredicate
+  | InfieldPredicate [Label] SimplePredicate
```
If the array is empty, this reduces to the other case, so you could either (a) simplify it to just `SearchTermSimple [Label] SimplePredicate` (which would represent all cases), or force the array to be non-empty by using either `Cofree Option Label` or `Tuple Label [Label]`.

jdegoes commented on src/Text/SlamSearch/Parser/Term.purs:
```diff
@@ -0,0 +1,202 @@
+    IncludeTerm t -> "Include(" <> show t <> ")"
+    ExcludeTerm t -> "Exclude(" <> show t <> ")"
+
+data PredicateAndLabel =
```
What do these mean / represent?

cryogenian replied:
There are four data structures representing steps of parsing:
- _Token_ - string parsed, basic tokens
- _Value_ - values constructed from tokens, operators pushed through for next stesp
- _PredicateAndLabel_ - labels moved to `L`, plus and minus to `I` for include and `E` for exclude. Operators and values composed to predicates.  
- _SearchTerm_ - labels bound to predicates. 

There is parser for each transition between states.

jdegoes commented on src/Text/SlamSearch/Parser/Term.purs:
```diff
@@ -0,0 +1,202 @@
+  case Tuple ls p of
+    Tuple [] (P p) -> return $ SimplePredicate p
+    Tuple xs (P p) -> return $ InfieldPredicate xs p
+    Tuple _ _ -> fail "impossible happens in simpleTerm"
```
Why is this impossible, and why can't we represent that in the type system?

jdegoes commented on src/Text/SlamSearch/Parser.purs:
```diff
@@ -0,0 +1,17 @@
+module Text.SlamSearch.Parser where
```
We should export only the needed functions.

cryogenian replied:
Do you mean completely hide everything except `tokens`, `value` and `search` in other modules?

jdegoes commented on src/Text/SlamSearch/Parser/Term.purs:
```diff
@@ -0,0 +1,202 @@
+likeP :: Parser [Value] SimplePredicate
+likeP = get (Through Tilde) *> (LikeP <$> when isVal) 
+  
+p :: Parser [Value] PredicateAndLabel
```
Letters like 'p', etc. don't convey much information, can we try for better names that describe what terms they parse?

jdegoes commented on src/Text/SlamSearch/Parser/Tokens.purs:
```diff
@@ -0,0 +1,194 @@
+rawString = do
+  cs <- many $ noneOf keyChars
+  if length cs == 0 then
+    fail "incorrect raw string"
```
This is a runtime failure, we should try to find a way to avoid it.

cryogenian replied:
It just returns `Left "incorrect raw string"` which stops this parser.

jdegoes commented on src/Text/SlamSearch/Parser/Tokens.purs:
```diff
@@ -0,0 +1,194 @@
+  between (string "\"") (string "\"") $ do
+    cs <- many quotedSymbol
+    if length cs == 0 then
+      fail "incorrect quoted string"
```
Ditto.

jdegoes commented on src/Text/SlamSearch/Parser/Utils.purs:
```diff
@@ -0,0 +1,24 @@
+module Text.SlamSearch.Parser.Utils where
```
Could we submit these in a PR to purescript-parsing?

jdegoes commented on src/Text/SlamSearch/Parser/Values.purs:
```diff
@@ -0,0 +1,155 @@
+import Text.SlamSearch.Parser.Tokens
+
+data Value =
+  Value String
```
This constructor is weirdly named because all of these tags are `Value`. Maybe `RawText` or `Text` would be better.

jdegoes commented on src/Text/SlamSearch/Parser/Values.purs:
```diff
@@ -0,0 +1,155 @@
+isGlob (Glob _) = true
+isGlob _ = false 
+
+isVal :: Value -> Boolean
```
Since all these are `Value`, is there a better name for this than `isVal`?

jdegoes commented on test/Test/S.purs:
```diff
@@ -0,0 +1,217 @@
+import Control.Alternative
+
+
+foreign import raise """
```
Can you use `purescript-exceptions` here?

jdegoes commented on test/Test/Main.purs:
```diff
@@ -0,0 +1,7 @@
+
+import Test.S
+
+main = do
```
It would be nice to get a StrongCheck test which can verify that any arbitrary search ADT can be pretty-printed to a search query which can be parsed to the same search ADT.

jdegoes commented:
@cryogenian 
1. The build is failing due to gulp not being installed.
2. Can you squash all commits down?
3. Let me know when you're ready to merge.

jdegoes commented:
Error correction cannot easily be done without building it into the parser combinators. I think that's more than we want to bite off for now. The question is: when something does NOT parse, what do we search for? For now, I think it's OK to change the background color to red, gray out the button, and prevent searching. Ultimately I'm not sure that's the best solution (what search engine prevents you from searching? None of them, they just interpret the search as best as possible).

cryogenian commented:
- Moved to `mandragora-bucket`
- fixed travis 
- renamed `Value`
- Don't know how to rename `PredicateAndLabel`
- Added `Printer`
- Added `StrongCheck` tests
- Made pr to `purescript-parsing`
- renamed `somethingP` to `somethingPredicate`
- Added `SearchAnd` (it just splits input by whitespaces, though)
- Don't know what runtime error can be caused by using `length cs` but switched to pattern matching
- Most of functions are private now (`parseSearchTerm` is public because it's easier to test than `parseSearchQuery`) 
- `SearchTermSimple` always use `[Label]`
